src/analyses/variable-sensitivity/abstract_environment.cpp

```
@@ -468,3 +499,211 @@ std::vector<abstract_object_pointert> eval_operands(
 
   return operands;
 }
+
+///////////
+abstract_value_pointert as_value(const abstract_object_pointert &obj)
+{
+  auto context_value =
+    std::dynamic_pointer_cast<const context_abstract_objectt>(obj);
+
+  return context_value
+           ? as_value(context_value->unwrap_context())
+           : std::dynamic_pointer_cast<const abstract_value_objectt>(obj);
+}
+
+bool is_value(const abstract_object_pointert &obj)
+{
+  return as_value(obj) != nullptr;
+}
+
+static auto inverse_operations =
+  std::map<irep_idt, irep_idt>{{ID_equal, ID_notequal},
+                               {ID_notequal, ID_equal},
+                               {ID_le, ID_gt},
+                               {ID_lt, ID_ge},
+                               {ID_ge, ID_lt},
+                               {ID_gt, ID_le}};
+
+exprt invert_expr(const exprt &expr)
+{
+  auto expr_id = expr.id();
+  auto inverse_operation = inverse_operations.find(expr_id);
+  if(inverse_operation == inverse_operations.end())
+    return nil_exprt();
```

There is a subtle precision loss here; if we have `if (A || B)`  then when we performing the transform for the `GOTO` instructions we will wind up assuming `not_exprt(or_exprt(A, B))` which will fail here rather than assuming `and_exprt(not_exprt(A), not_exprt(B))`.

You can either extend `invert_exprt` and get it to handle logical operations... or (which might be simpler) have an extra boolean flag argument for "we are assuming this to be ..." which starts as true.  `assume_not` will flip this (as will `assume_or` see below) and in the handlers for any atomic condition, you check the flag and if it is false, invert_exprt and recurse.

---

Having rewritten `assume_or` in terms of `assume_and`, I think this is taken care of. If we're assuming `not_exprt(or_exprt(A, B))` then that will be transformed to (more or less) `not_expr(not_expr(and_exprt(not_exprt(A), not_exprt(B))))`. It's a little long winded, but I think it will evaluate correctly.

---

Sounds good; thanks.